File: test/language/indexed_offsets/python/BoolIndexedOffsetArrayTest.py
import unittest
import zserio

from testutils import getZserioApi

class BoolIndexedOffsetArrayTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "indexed_offsets.zs").bool_indexed_offset_array

    def testBitSizeOf(self):
        createWrongOffsets = False
        boolIndexedOffsetArray = self._createBoolIndexedOffsetArray(createWrongOffsets)
        self.assertEqual(BOOL_INDEXED_OFFSET_ARRAY_BIT_SIZE, boolIndexedOffsetArray.bitSizeOf())

    def testBitSizeOfWithPosition(self):
        createWrongOffsets = False
        boolIndexedOffsetArray = self._createBoolIndexedOffsetArray(createWrongOffsets)
        bitPosition = 1
        self.assertEqual(BOOL_INDEXED_OFFSET_ARRAY_BIT_SIZE - bitPosition,
                         boolIndexedOffsetArray.bitSizeOf(bitPosition))

    def testInitializeOffsets(self):
        createWrongOffsets = True
        boolIndexedOffsetArray = self._createBoolIndexedOffsetArray(createWrongOffsets)
        bitPosition = 0
        self.assertEqual(BOOL_INDEXED_OFFSET_ARRAY_BIT_SIZE,
                         boolIndexedOffsetArray.initializeOffsets(bitPosition))
        self._checkBoolIndexedOffsetArray(boolIndexedOffsetArray)

    def testInitializeOffsetsWithPosition(self):
        createWrongOffsets = True
        boolIndexedOffsetArray = self._createBoolIndexedOffsetArray(createWrongOffsets)
        bitPosition = 9
        self.assertEqual(BOOL_INDEXED_OFFSET_ARRAY_BIT_SIZE + bitPosition - 1,
                         boolIndexedOffsetArray.initializeOffsets(bitPosition))

        offsetShift = 1
        self._checkOffsets(boolIndexedOffsetArray, offsetShift)

    def testRead(self):
        writeWrongOffsets = False
        writer = zserio.BitStreamWriter()
        BoolIndexedOffsetArrayTest._writeBoolIndexedOffsetArrayToStream(writer, writeWrongOffsets)
        reader = zserio.BitStreamReader(writer.getByteArray())
        boolIndexedOffsetArray = self.api.BoolIndexedOffsetArray()
        boolIndexedOffsetArray.read(reader)
        self._checkBoolIndexedOffsetArray(boolIndexedOffsetArray)

    def testReadWrongOffsets(self):
        writeWrongOffsets = True
        writer = zserio.BitStreamWriter()
        BoolIndexedOffsetArrayTest._writeBoolIndexedOffsetArrayToStream(writer, writeWrongOffsets)
        reader = zserio.BitStreamReader(writer.getByteArray())
        boolIndexedOffsetArray = self.api.BoolIndexedOffsetArray()
        with self.assertRaises(zserio.PythonRuntimeException):
            boolIndexedOffsetArray.read(reader)

    def testWrite(self):
        createWrongOffsets = True
        boolIndexedOffsetArray = self._createBoolIndexedOffsetArray(createWrongOffsets)
        writer = zserio.BitStreamWriter()
        boolIndexedOffsetArray.write(writer)
        self._checkBoolIndexedOffsetArray(boolIndexedOffsetArray)
        reader = zserio.BitStreamReader(writer.getByteArray())
        readBoolIndexedOffsetArray = self.api.BoolIndexedOffsetArray.fromReader(reader)
        self._checkBoolIndexedOffsetArray(readBoolIndexedOffsetArray)
        self.assertTrue(boolIndexedOffsetArray == readBoolIndexedOffsetArray)

    def testWriteWithPosition(self):
        createWrongOffsets = True
        boolIndexedOffsetArray = self._createBoolIndexedOffsetArray(createWrongOffsets)
        writer = zserio.BitStreamWriter()
        bitPosition = 8
        writer.writeBits(0, bitPosition)
        boolIndexedOffsetArray.write(writer)

        offsetShift = 1
        self._checkOffsets(boolIndexedOffsetArray, offsetShift)

    def testWriteWrongOffsets(self):
        createWrongOffsets = True
        boolIndexedOffsetArray = self._createBoolIndexedOffsetArray(createWrongOffsets)
        writer = zserio.BitStreamWriter()
        with self.assertRaises(zserio.PythonRuntimeException):
            boolIndexedOffsetArray.write(writer, callInitializeOffsets=False)

    @staticmethod
    def _writeBoolIndexedOffsetArrayToStream(writer, writeWrongOffsets):
        currentOffset = ELEMENT0_OFFSET
        for i in range(NUM_ELEMENTS):
            if writeWrongOffsets and i == NUM_ELEMENTS - 1:
                writer.writeBits(WRONG_OFFSET, 32)
            else:
                writer.writeBits(currentOffset, 32)
            currentOffset += ALIGNED_ELEMENT_BYTE_SIZE

        writer.writeBits(SPACER_VALUE, 1)

        writer.writeBits(0, 7)
        for i in range(NUM_ELEMENTS):
            writer.writeBits((i & 0x01) != 0, ELEMENT_SIZE)
            if i != NUM_ELEMENTS - 1:
                writer.writeBits(0, ALIGNED_ELEMENT_SIZE - ELEMENT_SIZE)

    def _checkOffsets(self, boolIndexedOffsetArray, offsetShift):
        offsets = boolIndexedOffsetArray.getOffsets()
        self.assertEqual(NUM_ELEMENTS, len(offsets))
        expectedOffset = ELEMENT0_OFFSET + offsetShift
        for offset in offsets:
            self.assertEqual(expectedOffset, offset)
            expectedOffset += ALIGNED_ELEMENT_BYTE_SIZE

    def _checkBoolIndexedOffsetArray(self, boolIndexedOffsetArray):
        offsetShift = 0
        self._checkOffsets(boolIndexedOffsetArray, offsetShift)

        self.assertEqual(SPACER_VALUE, boolIndexedOffsetArray.getSpacer())

        data = boolIndexedOffsetArray.getData()
        self.assertEqual(NUM_ELEMENTS, len(data))
        for i in range(NUM_ELEMENTS):
            self.assertEqual((i & 0x01) != 0, data[i])

    def _createBoolIndexedOffsetArray(self, createWrongOffsets):
        offsets = []
        currentOffset = ELEMENT0_OFFSET
        for i in range(NUM_ELEMENTS):
            if createWrongOffsets and i == NUM_ELEMENTS - 1:
                offsets.append(WRONG_OFFSET)
            else:
                offsets.append(currentOffset)
            currentOffset += ALIGNED_ELEMENT_BYTE_SIZE

        data = []
        for i in range(NUM_ELEMENTS):
            data.append((i & 0x01) != 0)

        return self.api.BoolIndexedOffsetArray(offsets, SPACER_VALUE, data)

NUM_ELEMENTS = 5

WRONG_OFFSET = 0

ELEMENT0_OFFSET = NUM_ELEMENTS * 4 + 1
ELEMENT_SIZE = 1
ALIGNED_ELEMENT_SIZE = 8
ALIGNED_ELEMENT_BYTE_SIZE = 1

SPACER_VALUE = 1

BOOL_INDEXED_OFFSET_ARRAY_BIT_SIZE = (ELEMENT0_OFFSET * 8 + (NUM_ELEMENTS - 1) * ALIGNED_ELEMENT_SIZE +
                                      ELEMENT_SIZE)

File: test/language/indexed_offsets/cpp/BoolIndexedOffsetArrayTest.cpp
#include "gtest/gtest.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

#include "indexed_offsets/bool_indexed_offset_array/BoolIndexedOffsetArray.h"

namespace indexed_offsets
{
namespace bool_indexed_offset_array
{

class BoolIndexedOffsetArrayTest : public ::testing::Test
{
protected:
    void writeBoolIndexedOffsetArrayToByteArray(zserio::BitStreamWriter& writer, bool writeWrongOffsets)
    {
        const uint32_t wrongOffset = WRONG_OFFSET;
        uint32_t currentOffset = ELEMENT0_OFFSET;
        for (uint8_t i = 0; i < NUM_ELEMENTS; ++i)
        {
            if ((i + 1) == NUM_ELEMENTS && writeWrongOffsets)
                writer.writeBits(wrongOffset, 32);
            else
                writer.writeBits(currentOffset, 32);
            currentOffset += ALIGNED_ELEMENT_BYTE_SIZE;
        }

        writer.writeBits(SPACER_VALUE, 1);
        writer.writeBits(0, 7);

        for (uint8_t i = 0; i < NUM_ELEMENTS; ++i)
        {
            writer.writeBits(i & 0x01, ELEMENT_SIZE);
            if ((i + 1) != NUM_ELEMENTS)
                writer.writeBits(0, ALIGNED_ELEMENT_SIZE - ELEMENT_SIZE);
        }
    }

    void checkOffsets(const BoolIndexedOffsetArray& boolIndexedOffsetArray, uint16_t offsetShift)
    {
        const auto& offsets = boolIndexedOffsetArray.getOffsets();
        const size_t expectedNumElements = NUM_ELEMENTS;
        ASSERT_EQ(expectedNumElements, offsets.size());
        uint32_t expectedOffset = ELEMENT0_OFFSET + offsetShift;
        for (auto it = offsets.begin(); it != offsets.end(); ++it)
        {
            ASSERT_EQ(expectedOffset, *it);
            expectedOffset += ALIGNED_ELEMENT_BYTE_SIZE;
        }
    }

    void checkBoolIndexedOffsetArray(const BoolIndexedOffsetArray& boolIndexedOffsetArray)
    {
        const uint16_t offsetShift = 0;
        checkOffsets(boolIndexedOffsetArray, offsetShift);

        const uint8_t expectedSpacer = SPACER_VALUE;
        ASSERT_EQ(expectedSpacer, boolIndexedOffsetArray.getSpacer());

        const auto& data = boolIndexedOffsetArray.getData();
        const size_t expectedNumElements = NUM_ELEMENTS;
        ASSERT_EQ(expectedNumElements, data.size());
        for (uint8_t i = 0; i < NUM_ELEMENTS; ++i)
            ASSERT_EQ((i & 0x01) ? true : false, data[i] != 0);
    }

    void fillBoolIndexedOffsetArray(BoolIndexedOffsetArray& boolIndexedOffsetArray, bool createWrongOffsets)
    {
        auto& offsets = boolIndexedOffsetArray.getOffsets();
        offsets.reserve(NUM_ELEMENTS);
        const uint32_t wrongOffset = WRONG_OFFSET;
        uint32_t currentOffset = ELEMENT0_OFFSET;
        for (uint8_t i = 0; i < NUM_ELEMENTS; ++i)
        {
            if ((i + 1) == NUM_ELEMENTS && createWrongOffsets)
                offsets.push_back(wrongOffset);
            else
                offsets.push_back(currentOffset);
            currentOffset += ALIGNED_ELEMENT_BYTE_SIZE;
        }
        boolIndexedOffsetArray.setSpacer(SPACER_VALUE);

        auto& data = boolIndexedOffsetArray.getData();
        data.reserve(NUM_ELEMENTS);
        for (uint8_t i = 0; i < NUM_ELEMENTS; ++i)
            data.push_back((i & 0x01) ? true : false);
    }

    static const uint8_t    NUM_ELEMENTS = 5;

    static const uint32_t   WRONG_OFFSET = 0;

    static const uint32_t   ELEMENT0_OFFSET = NUM_ELEMENTS * sizeof(uint32_t) + sizeof(uint8_t);
    static const uint8_t    ELEMENT_SIZE = 1;
    static const uint8_t    ALIGNED_ELEMENT_SIZE = 8;
    static const uint8_t    ALIGNED_ELEMENT_BYTE_SIZE = ALIGNED_ELEMENT_SIZE / 8;

    static const uint8_t    SPACER_VALUE = 1;

    static const size_t     BOOL_indexed_offset_ARRAY_BIT_SIZE = ELEMENT0_OFFSET * 8 +
            (NUM_ELEMENTS - 1) * ALIGNED_ELEMENT_SIZE + ELEMENT_SIZE;

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
};

TEST_F(BoolIndexedOffsetArrayTest, read)
{
    const bool writeWrongOffsets = false;
    zserio::BitStreamWriter writer(bitBuffer);
    writeBoolIndexedOffsetArrayToByteArray(writer, writeWrongOffsets);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    BoolIndexedOffsetArray boolIndexedOffsetArray(reader);
    checkBoolIndexedOffsetArray(boolIndexedOffsetArray);
}

TEST_F(BoolIndexedOffsetArrayTest, readWrongOffsets)
{
    const bool writeWrongOffsets = true;
    zserio::BitStreamWriter writer(bitBuffer);
    writeBoolIndexedOffsetArrayToByteArray(writer, writeWrongOffsets);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    EXPECT_THROW(BoolIndexedOffsetArray boolIndexedOffsetArray(reader), zserio::CppRuntimeException);
}

TEST_F(BoolIndexedOffsetArrayTest, bitSizeOf)
{
    const bool createWrongOffsets = false;
    BoolIndexedOffsetArray boolIndexedOffsetArray;
    fillBoolIndexedOffsetArray(boolIndexedOffsetArray, createWrongOffsets);

    const size_t expectedBitSize = BOOL_indexed_offset_ARRAY_BIT_SIZE;
    ASSERT_EQ(expectedBitSize, boolIndexedOffsetArray.bitSizeOf());
}

TEST_F(BoolIndexedOffsetArrayTest, bitSizeOfWithPosition)
{
    const bool createWrongOffsets = false;
    BoolIndexedOffsetArray boolIndexedOffsetArray;
    fillBoolIndexedOffsetArray(boolIndexedOffsetArray, createWrongOffsets);

    const size_t bitPosition = 1;
    const size_t expectedBitSize = BOOL_indexed_offset_ARRAY_BIT_SIZE - bitPosition;
    ASSERT_EQ(expectedBitSize, boolIndexedOffsetArray.bitSizeOf(bitPosition));
}

TEST_F(BoolIndexedOffsetArrayTest, initializeOffsets)
{
    const bool createWrongOffsets = true;
    BoolIndexedOffsetArray boolIndexedOffsetArray;
    fillBoolIndexedOffsetArray(boolIndexedOffsetArray, createWrongOffsets);

    const size_t bitPosition = 0;
    const size_t expectedBitSize = BOOL_indexed_offset_ARRAY_BIT_SIZE;
    ASSERT_EQ(expectedBitSize, boolIndexedOffsetArray.initializeOffsets(bitPosition));
    checkBoolIndexedOffsetArray(boolIndexedOffsetArray);
}

TEST_F(BoolIndexedOffsetArrayTest, initializeOffsetsWithPosition)
{
    const bool createWrongOffsets = true;
    BoolIndexedOffsetArray boolIndexedOffsetArray;
    fillBoolIndexedOffsetArray(boolIndexedOffsetArray, createWrongOffsets);

    const size_t bitPosition = 9;
    const size_t expectedBitSize = BOOL_indexed_offset_ARRAY_BIT_SIZE + bitPosition - 1;
    ASSERT_EQ(expectedBitSize, boolIndexedOffsetArray.initializeOffsets(bitPosition));

    const uint16_t offsetShift = 1;
    checkOffsets(boolIndexedOffsetArray, offsetShift);
}

TEST_F(BoolIndexedOffsetArrayTest, write)
{
    const bool createWrongOffsets = true;
    BoolIndexedOffsetArray boolIndexedOffsetArray;
    fillBoolIndexedOffsetArray(boolIndexedOffsetArray, createWrongOffsets);

    zserio::BitStreamWriter writer(bitBuffer);
    boolIndexedOffsetArray.write(writer);
    checkBoolIndexedOffsetArray(boolIndexedOffsetArray);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    BoolIndexedOffsetArray readBoolIndexedOffsetArray(reader);
    checkBoolIndexedOffsetArray(readBoolIndexedOffsetArray);
    ASSERT_TRUE(boolIndexedOffsetArray == readBoolIndexedOffsetArray);
}

TEST_F(BoolIndexedOffsetArrayTest, writeWithPosition)
{
    const bool createWrongOffsets = true;
    BoolIndexedOffsetArray boolIndexedOffsetArray;
    fillBoolIndexedOffsetArray(boolIndexedOffsetArray, createWrongOffsets);

    zserio::BitStreamWriter writer(bitBuffer);
    const size_t bitPosition = 8;
    writer.writeBits(0, bitPosition);
    boolIndexedOffsetArray.write(writer);

    const uint16_t offsetShift = 1;
    checkOffsets(boolIndexedOffsetArray, offsetShift);
}

TEST_F(BoolIndexedOffsetArrayTest, writeWrongOffsets)
{
    const bool createWrongOffsets = true;
    BoolIndexedOffsetArray boolIndexedOffsetArray;
    fillBoolIndexedOffsetArray(boolIndexedOffsetArray, createWrongOffsets);

    zserio::BitStreamWriter writer(bitBuffer);
    ASSERT_THROW(boolIndexedOffsetArray.write(writer, zserio::NO_PRE_WRITE_ACTION),
            zserio::CppRuntimeException);
}

} // namespace bool_indexed_offset_array
} // namespace indexed_offsets
